### PR TITLE
implement epoch->cid on postgres so it doesn't jank-index

### PR DESCRIPTION
**Pull request by willscott**

### CODE REVIEW

**Comment by willscott** [comment]
the extra methods mean that the postgres datastore should implement lib.datastore, which means it'll skip indexing https://github.com/filecoin-project/statediff/blob/master/lib/lazysource.go#L163

File: lib/tstracker/chainbs.go [hunk 1]
```diff
@@ -129,3 +132,47 @@ func NewTrackingPgChainstore(ctx context.Context, conn string) (TrackingChainsto
...
+	tsk, _, err := t.CurrentDbTipSetKey(ctx)
+	if err != nil {
+		return cid.Undef
+	}
+	cids := tsk.Cids()
+	if len(cids) > 0 {
+		return cids[0]
+	}
+	return cid.Undef
+}
```
Comment (ribasushi) [inline lib/tstracker/chainbs.go:157]: Is this intended/expected? The amount of times we end up with a single-block tipset is *very* rare

File: lib/tstracker/chainbs.go [hunk 2]
```diff
@@ -129,3 +132,47 @@ func NewTrackingPgChainstore(ctx context.Context, conn string) (TrackingChainsto
...
+func (t *tcs) CidAtHeight(ctx context.Context, h int64) (cid.Cid, error) {
+	tsd, err := t.GetFilTipSetHead(ctx)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	tsd, err = t.FindFilTipSet(ctx, tsd.TipSetCids, abi.ChainEpoch(h))
```
Comment (ribasushi) [inline lib/tstracker/chainbs.go:164]: This is incorrect - you need to specify a walkback, something like ( untested )
```suggestion
	tsd, err = t.FindFilTipSet(ctx, tsd.TipSetCids, tsd.Epoch - abi.ChainEpoch(h))
```

**Review by ribasushi — CHANGES_REQUESTED** [review]
LGTM as a whole - has a mistake though

File: lib/tstracker/chainbs.go [hunk 3]
```diff
@@ -129,3 +132,47 @@ func NewTrackingPgChainstore(ctx context.Context, conn string) (TrackingChainsto
...
 	return &tcs{PgBlockstore: pgbs}, nil
 }
+
+func (*tcs) DeleteMany([]cid.Cid) error {
+	return errors.New("not supported")
+}
+
+func (*tcs) View(cid.Cid, func([]byte) error) error {
+	return errors.New("not supported")
+}
```
Comment (ribasushi) [inline lib/tstracker/chainbs.go:142]: I do implement these btw:
https://github.com/filecoin-project/go-bs-postgres-chainnotated/blob/feat/the_force_awkens/store.go#L71-L73

File: lib/tstracker/chainbs.go (hunk 1 above)
Comment (ribasushi) [reply lib/tstracker/chainbs.go:157]: Nevermind, conusion resolved

**Review by ribasushi — APPROVED** [review]
:shipit:

**Pull request merged**